### PR TITLE
Migrate from compilerflasher to using codebender's builder service

### PR DESCRIPTION
**Pull request by kyleoliveira**
Working toward resolving issue #6 . Not complete yet, but it's a good start.

Note that I've branched off of the server_quickstart branch rather than master because the pull request for that hasn't been merged in to master yet.

### CODE REVIEW

**Comment by spiliot** [comment]
I've added the hold flag since the branch is wip. Remove it when the code is considered ready to merge.

Also you will need to rebase to master since `server_quickstart` branch is already merged and some commits were squashed.

**Comment by kyleoliveira** [comment]
Sounds good. Thanks!

**Comment by kyleoliveira** [comment]
Just FYI, I've squashed the previous commits into a single commit ( 4d63fdcd73b016687498fcacb468fbcb6221d55a ).

**Comment by kyleoliveira** [comment]
I think we're ready to go with this. I've deployed the most recent code to the blockly.codebender.cc site for you to try out (we may want to have something like dev.blockly.codebender.cc sometime in the future...).

Please let me know if it looks good and I'll squash it into a single commit.

**Comment by tzikis** [comment]
Hey Kyle,
From my part, I am on an iPad and can't actually check it out and see if it works, but just as my input, it looks like you guys have resized the actual blockly view and made it really small and unusable, so it's a bit hard for us to show it to people. Any chance you could fix that?

File: app/Blocklyduino.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+        // The URL to use to connect to the Codebender Builder service
+        $this['builder_url'] = "http://builder.codebender.cc:8080/compile?api_key=blocklyduino";
+
+        // An example JSON request for testing the service connection
+        $this['debug_code_request'] = <<<HDOC
+{"files":[{"filename":"Blink Example.ino","content":"\/*\\n\\tBlink\\n\\tTurns on an LED on for one second, then off for one second, repeatedly.\\n\\n\\tThis example code is in the public domain.\\n*\/\\n\\nvoid setup()\\n{\\n\\t\/\/ initialize the digital pin as an output.\\n\\t\/\/ Pin 13 has an LED connected on most Arduino boards:\\n\\tpinMode(13, OUTPUT);\\n}\\n\\nvoid loop()\\n{\\n\\tdigitalWrite(13, HIGH); \/\/ set the LED on\\n\\tdelay(1000); \/\/ wait for a second\\n\\tdigitalWrite(13, LOW); \/\/ set the LED off\\n\\tdelay(1000); \/\/ wait for a second\\n}\\n"}],"libraries":[],"logging":true,"format":"binary","version":"105","build":{"mcu":"atmega328p","f_cpu":"16000000L","core":"arduino","variant":"standard"}}
+HDOC;
+
+        foreach ($values as $key => $value) {
```
Comment (spiliot) [inline app/Blocklyduino.php:35]: I think what you're doing here is exactly what [php's `extract`](http://php.net/manual/en/function.extract.php) does (probably in a more efficient way)

File: public/assets/scripts/blocklyflasher.js [hunk 2]
```diff
@@ -0,0 +1,1790 @@
...
+                "tabID": this.tabID
+            };
+            createLogFlasher(actionId, metaData);
+        };
+
+        this.oldPort = null;
+
+        this.logPorts = function () {
+            var oldPort;
+            if (this.oldPort == null)
```
Comment (spiliot) [inline public/assets/scripts/blocklyflasher.js]: It's good practice (actually very very good practice!) to always include curly brackets even if it is a one liner for `if`, `for`, `switch`, etc. It doesn't cost anything and can save hours of crazy debugging.

File: public/assets/scripts/blocklyflasher.js [hunk 3]
```diff
@@ -0,0 +1,1790 @@
...
+                    "tabID": this.tabID
+                };
+                createLogFlasher(actionId, metaData);
+            }
+        };
+
+        this.savePort = function() {
+            var cb = this;
+            if(typeof Lawnchair !== "undefined")
+            {
```
Comment (spiliot) [inline public/assets/scripts/blocklyflasher.js]: It's always best to be consistent with your coding style. In this document you have `if`s defined as:
```js
if (something) {
    // do something
}```
,
```js
if (somethingElse){ //note no space after )
//or this
if(somethingElse) { //note no space between if and (
```
and 
```js
if (another)
{
    //something
}
```

Of all those I suggest sticking with the first, it is visually consistent with PHP's PSR-2 coding style which means if you switch between the two it won't be mentally confusing at all.

File: public/assets/scripts/blocklyflasher.js [hunk 4]
```diff
@@ -0,0 +1,1790 @@
...
+        };
+
+        this.loadPort = function() {
+            var cb = this;
+            if (this.currentPorts.length > 0) {
+                $('#cb_cf_ports').find('option[value=""]').remove();
+                if(typeof Lawnchair !== "undefined") {
+                    Lawnchair(function () {
+                        this.exists('port', function (exists) {
+                            if (exists) {
```
Comment (spiliot) [inline public/assets/scripts/blocklyflasher.js]: By this line I'm already looking for paper to note what each indent is and what if/elses do. 

There is a school that says to
1. Not have more than one indent level 
2. Never use else

Both can work 99% of the time and once you adopt them code looks much nicer and cleaner. Look up the subject of code calisthenics, they come from OOP but can be applied to most languages for most part.

**Comment by kyleoliveira** [comment]
With respect to the sizing:
The asset path was returning localhost for the base url. I missed that in the transition over to the AWS machine somehow (it was like 1AM, so maybe that had something to do with it). I've fixed that on the AWS machine so that the asset path is returned as a relative path and it seems happy now. I'll add it in a followup commit.

File: app/Blocklyduino.php (hunk 1 above)
Comment (kyleoliveira) [inline app/Blocklyduino.php:35]: I'm not sure about efficiency, but I was trying to follow convention. All the Silex tutorials I saw assigned Application configuration in this sort of way. In fact, if you checkout the code for Silex\Application (https://github.com/silexphp/Silex/blob/1.2/src/Silex/Application.php), you'll see they do something similar.

**Comment by kyleoliveira** [comment]
Please use #11 instead. That branch is a clean, squashed-down version of this one.

**Comment by spiliot** [comment]
Closing as #11 is now relevant

**Pull request closed without merging**